### PR TITLE
support infinite loop over alpaca dataset

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #66

Previously, alpaca dataset is consumed up after only ~50 iterations with 8 data parallel ranks and 8 batch size. This PR adds the (default) option to loop infinitely on the dataset, so that we can unblock integrating other functionalities. Note that loss-related metrics should be read with caution as this will cause overfit.

Update: moved to #92 because migrating to pytorch/ confused ghstack.

### CODE REVIEW

File: torchtrain/datasets/alpaca.py [hunk 1]
```diff
@@ -43,38 +44,48 @@ def __init__(
...
+                while len(all_tokens) >= max_buffer_token_len:
+                    x = torch.LongTensor(all_tokens[:max_buffer_token_len])
+                    # batched_x = x.reshape(self.batch_size, -1)
+                    # update tokens to the remaining tokens
+                    all_tokens = all_tokens[max_buffer_token_len:]
+                    input = x[:-1]
+                    label = x[1:]
+                    yield input, label
+            if not self.infinite:
+                break
```
Comment (XilunWu) [inline torchtrain/datasets/alpaca.py:78]: should we add some mechanic to allow a stop? `self.infinite` is a constant after being initialized.

File: torchtrain/datasets/alpaca.py [hunk 2]
```diff
@@ -43,38 +44,48 @@ def __init__(
...
-                x = torch.LongTensor(all_tokens[:max_buffer_token_len])
-                # batched_x = x.reshape(self.batch_size, -1)
-                # update tokens to the remaining tokens
-                all_tokens = all_tokens[max_buffer_token_len:]
-                input = x[:-1]
-                label = x[1:]
-                yield input, label
+                while len(all_tokens) >= max_buffer_token_len:
+                    x = torch.LongTensor(all_tokens[:max_buffer_token_len])
+                    # batched_x = x.reshape(self.batch_size, -1)
```
Comment (wanchaol) [inline torchtrain/datasets/alpaca.py:71]: nit: we can delete the staled comment?

File: torchtrain/datasets/alpaca.py (hunk 1 above)
Comment (wanchaol) [reply torchtrain/datasets/alpaca.py:78]: I think cmd + c should be sufficient?
